### PR TITLE
[mtv] Added support for MTV Italia (mtv.it)

### PR DESCRIPTION
**Pull request by nixxo**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
**Added support for MTV Italia (mtv.it)**
- MTVItalia: Support for single video
- MTVItaliaProgramma: Support for series/playlist videos
- MTVServicesInfoExtractor: Added series/season_number/episode_number tag extraction

### CODE REVIEW

File: yt_dlp/extractor/mtv.py [hunk 1]
```diff
@@ -483,3 +503,156 @@ def _get_feed_query(self, uri):
...
+                    yield self.url_result(entry['canonicalURL'])
+            if not url:
+                break
+
+    def _real_extract(self, url):
+        query = {'url': url}
+        info_url = update_url_query(self._FEED_URL, query)
+        video_id = self._match_id(url)
+        info = self._download_json(info_url, video_id)
+        info = info.get('manifest')
```
Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:600]:
```suggestion
        info = self._download_json(info_url, video_id).get('manifest')
```

Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:600]:
```suggestion
        info = self._download_json(info_url, video_id).get('manifest')
```

File: yt_dlp/extractor/mtv.py [hunk 2]
```diff
@@ -483,3 +503,156 @@ def _get_feed_query(self, uri):
...
+                lambda x: x['title'],
+                lambda x: x['headline']),
+            compat_str)
+        description = try_get(info, lambda x: x['content'], compat_str)
+
+        if current_url:
+            season = self._download_json(
+                playlist_url, video_id, 'Seasons infos')
+            season = try_get(
+                season, lambda x: x['result']['data'], dict)
```
Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:643]:
```suggestion
            season = try_get(
                self._download_json(playlist_url, video_id, 'Seasons infos'),
                lambda x: x['result']['data'], dict)
```

**Comment by pukkandan** [comment]
* I cannot test this website, so please confirm that all the tests work correctly
* If this was a PR from youtube-dl, please link to it for my reference

**Comment by nixxo** [comment]
4 tests (1 for MTVItalia, 3 for MTVItaliaProgramma)

```
λ python test\test_download.py TestDownload.test_MTVItalia                                                                                                                    
[debug] Loading archive file None                                                                                                                                             
[debug] Using fake IP 79.25.8.35 (IT) as X-Forwarded-For.                                                                                                                     
[mtv.it] mario-una-serie-di-maccio-capatonda-cavoli-amario-episodio-completo-S1-E1: Downloading webpage                                                                       
[mtv.it] Downloading JSON metadata                                                                                                                                            
[mtv.it] 2e0c4723-1ad5-4218-92e7-6a07196337e7: Downloading info                                                                                                               
[mtv.it] 0f0fc78e-45fc-4cce-8f24-971c25477530: Extracting information                                                                                                         
[mtv.it] 0f0fc78e-45fc-4cce-8f24-971c25477530: Downloading video urls                                                                                                         
[mtv.it] 0f0fc78e-45fc-4cce-8f24-971c25477530: Downloading m3u8 information                                                                                                   
[debug] Sort order given by user:                                                                                                                                             
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id 
[debug] Sort order given by user:                                                                                                                                             
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id 
[download] Downloading playlist: Cavoli amario (episodio completo)                                                                                                            
[info] Writing playlist metadata as JSON to: test_MTVItalia_NA.info.json                                                                                                      
[mtv.it] playlist Cavoli amario (episodio completo): Collected 1 video ids (downloading 1 of them)                                                                            
[download] Downloading video 1 of 1                                                                                                                                           
[info] Downloading format(s) hls-4716                                                                                                                                         
[info] Writing video metadata as JSON to: test_MTVItalia_0f0fc78e-45fc-4cce-8f24-971c25477530.info.json                                                                       
[download] Finished downloading playlist: Cavoli amario (episodio completo)                                                                                                   
.                                                                                                                                                                             
----------------------------------------------------------------------                                                                                                        
Ran 1 test in 4.927s                                                                                                                                                          
                                                                                                                                                                              
OK
```
                                                                                                                            
```
λ python test\test_download.py TestDownload.test_MTVItaliaProgramma                                                                                                           
[debug] Loading archive file None                                                                                                                                             
[debug] Using fake IP 79.19.210.179 (IT) as X-Forwarded-For.                                                                                                                  
[mtv.it:programma] s2rppv: Downloading JSON metadata                                                                                                                          
[mtv.it:programma] a6f155bc-8220-4640-aa43-9b95f64ffa3d: Show infos                                                                                                           
[mtv.it:programma] a6f155bc-8220-4640-aa43-9b95f64ffa3d: Seasons infos                                                                                                        
[download] Downloading playlist: Mario - Una Serie Di Maccio Capatonda                                                                                                        
[info] Writing playlist metadata as JSON to: test_MTVItaliaProgramma_a6f155bc-8220-4640-aa43-9b95f64ffa3d.info.json                                                           
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda: page 1                                                                                                              
[mtv.it:programma] playlist Mario - Una Serie Di Maccio Capatonda: Downloading 2 videos                                                                                       
[download] Downloading video 1 of 2                                                                                                                                           
[download] Downloading video 2 of 2                                                                                                                                           
[download] Finished downloading playlist: Mario - Una Serie Di Maccio Capatonda                                                                                               
.                                                                                                                                                                             
----------------------------------------------------------------------                                                                                                        
Ran 1 test in 0.750s                                                                                                                                                          
                                                                                                                                                                              
OK
```                                                                                                                               
```
λ python test\test_download.py TestDownload.test_MTVItaliaProgramma_1                                                                                                         
[debug] Loading archive file None                                                                                                                                             
[debug] Using fake IP 79.59.23.179 (IT) as X-Forwarded-For.                                                                                                                   
[mtv.it:programma] d9ncjf: Downloading JSON metadata                                                                                                                          
[mtv.it:programma] 4deeb5d8-f272-490c-bde2-ff8d261c6dd1: Show infos                                                                                                           
[mtv.it:programma] 4deeb5d8-f272-490c-bde2-ff8d261c6dd1: Seasons infos                                                                                                        
[download] Downloading playlist: Mario - Una Serie Di Maccio Capatonda - Stagione 2                                                                                           
[info] Writing playlist metadata as JSON to: test_MTVItaliaProgramma_1_4deeb5d8-f272-490c-bde2-ff8d261c6dd1.info.json                                                         
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda - Stagione 2: page 1                                                                                                 
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda - Stagione 2: page 2                                                                                                 
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda - Stagione 2: page 3                                                                                                 
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda - Stagione 2: page 4                                                                                                 
[mtv.it:programma] Mario - Una Serie Di Maccio Capatonda - Stagione 2: page 5                                                                                                 
[mtv.it:programma] playlist Mario - Una Serie Di Maccio Capatonda - Stagione 2: Downloading 34 videos                                                                         
[download] Downloading video 1 of 34                                                                                                                                          
[download] Downloading video 2 of 34                                                                                                                                          
...                                                                                                                                    
[download] Downloading video 33 of 34                                                                                                                                         
[download] Downloading video 34 of 34                                                                                                                                         
[download] Finished downloading playlist: Mario - Una Serie Di Maccio Capatonda - Stagione 2                                                                                  
.                                                                                                                                                                             
----------------------------------------------------------------------                                                                                                        
Ran 1 test in 5.160s                                                                                                                                                          
                                                                                                                                                                              
OK
```                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                      
```
λ python test\test_download.py TestDownload.test_MTVItaliaProgramma_2                                                                                                         
[debug] Loading archive file None                                                                                                                                             
[debug] Using fake IP 79.53.139.129 (IT) as X-Forwarded-For.                                                                                                                  
[mtv.it:programma] sexy: Downloading JSON metadata                                                                                                                            
[mtv.it:programma] sexy: Downloading JSON metadata                                                                                                                            
[mtv.it:programma] dee8f9ee-756d-493b-bf37-16d1d2783359: Show infos                                                                                                           
[download] Downloading playlist: Sexy Videos                                                                                                                                  
[info] Writing playlist metadata as JSON to: test_MTVItaliaProgramma_2_dee8f9ee-756d-493b-bf37-16d1d2783359.info.json                                                         
[mtv.it:programma] Sexy Videos: page 1                                                                                                                                        
[mtv.it:programma] playlist Sexy Videos: Downloading 148 videos                                                                                                               
[download] Downloading video 1 of 148                                                                                                                                         
[download] Downloading video 2 of 148                                                                                                                                         
...                                                                                                                                  
[download] Downloading video 147 of 148                                                                                                                                       
[download] Downloading video 148 of 148                                                                                                                                       
[download] Finished downloading playlist: Sexy Videos                                                                                                                         
.                                                                                                                                                                             
----------------------------------------------------------------------                                                                                                        
Ran 1 test in 1.527s                                                                                                                                                          
                                                                                                                                                                              
OK
```

File: yt_dlp/extractor/mtv.py [hunk 3]
```diff
@@ -483,3 +503,154 @@ def _get_feed_query(self, uri):
...
+    def _real_extract(self, url):
+        query = {'url': url}
+        info_url = update_url_query(self._FEED_URL, query)
+        video_id = self._match_id(url)
+        info = self._download_json(info_url, video_id).get('manifest')
+
+        redirect = try_get(
+            info, lambda x: x['newLocation']['url'], compat_str)
+        if redirect:
+            return self.url_result(redirect)
```
Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:604]: Check whether `url` or `url_transparent` is more suitable

File: yt_dlp/extractor/mtv.py [hunk 4]
```diff
@@ -483,3 +503,154 @@ def _get_feed_query(self, uri):
...
+
+        playlist_url = current_url = None
+        zones = info.get('zones') or []
+        for z in zones:
+            if zones[z].get('moduleName') in {'INTL_M304', 'INTL_M209'}:
+                info_url = zones[z].get('feed')
+            if zones[z].get('moduleName') in {'INTL_M308', 'INTL_M317'}:
+                playlist_url = playlist_url or zones[z].get('feed')
+            if zones[z].get('moduleName') in {'INTL_M300'}:
+                current_url = current_url or zones[z].get('feed')
```
Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:620]: Untested!

```suggestion
        for z in (info.get('zones') or {}).values():
            if z.get('moduleName') in ('INTL_M304', 'INTL_M209'):
                info_url = z.get('feed')
            if z.get('moduleName') in ('INTL_M308', 'INTL_M317'):
                playlist_url = playlist_url or z.get('feed')
            if z.get('moduleName') in ('INTL_M300',):
                current_url = current_url or z.get('feed')
```

File: yt_dlp/extractor/mtv.py [hunk 5]
```diff
@@ -483,3 +503,154 @@ def _get_feed_query(self, uri):
...
+        title = title or try_get(
+            info, (
+                lambda x: x['title'],
+                lambda x: x['headline']),
+            compat_str)
+        description = try_get(info, lambda x: x['content'], compat_str)
+
+        if current_url:
+            season = try_get(
+                self._download_json(playlist_url, video_id, 'Seasons infos'),
```
Comment (pukkandan) [inline yt_dlp/extractor/mtv.py:640]:
```suggestion
                self._download_json(playlist_url, video_id, 'Seasons info'),
```

File: yt_dlp/extractor/mtv.py [hunk 6]
```diff
@@ -483,3 +503,154 @@ def _get_feed_query(self, uri):
...
+            'skip_download': True,
+        },
+    }]
+    _GEO_COUNTRIES = ['IT']
+    _FEED_URL = 'http://www.mtv.it/feeds/triforce/manifest/v8'
+
+    def _get_entries(self, title, url):
+        while True:
+            pg = re.search(r'/(\d+)$', url)
+            pg = pg.group(1) if pg else '1'
```
[the remark this answers is not in the document]
Comment (nixxo) [reply yt_dlp/extractor/mtv.py:580]: 1. the url is extracted from a json in the first place and does not contain any query; it adds /1 /2 etc. for every page of data
2. ok. I'll do that.

File: yt_dlp/extractor/mtv.py (hunk 3 above)
Comment (nixxo) [reply yt_dlp/extractor/mtv.py:604]: no need for url_transparent

**Comment by pukkandan** [comment]
hey @nixxo, unrelated to this PR, but I was wondering if you are on the discord server

**Comment by nixxo** [comment]
not yet

**Pull request merged**